zebra/zebra_vty.c

```diff
@@ -433,13 +436,27 @@ static void zebra_show_ip_route_opaque(struct vty *vty, struct route_entry *re,
 			vty_out(vty, "    Opaque Data: %s",
 				(char *)re->opaque->data);
 		break;
-	case ZEBRA_ROUTE_BGP:
-		if (json)
-			json_object_string_add(json, "asPath",
-					       (char *)re->opaque->data);
-		else
-			vty_out(vty, "    AS-Path: %s",
-				(char *)re->opaque->data);
+	case ZEBRA_ROUTE_BGP: {
+		memcpy(&bzo, re->opaque->data, sizeof(bzo));
```

Comment:
In `zebra/zapi_msg.c` I see this:

```c
	if (CHECK_FLAG(api.message, ZAPI_MESSAGE_OPAQUE)) {
		re->opaque = XMALLOC(MTYPE_OPAQUE,
				     sizeof(struct opaque) + api.opaque.length);
		re->opaque->length = api.opaque.length;
		memcpy(re->opaque->data, api.opaque.data, re->opaque->length);
	}
```

Is it possible that you might be copying data outside the allocated memory area?

Reply:
You mean `api.opaque.data` > `re->opaque->data` or what actually?

Reply:
Sorry, don't understand this yet. BGP does not know anything about re (route_entry struct).

Reply:
But re->opaque->data can be anything, not only `struct bgp_zebra_opaque`. 

An assertion is in the decode part (zapi_route_decode):

```c
	if (CHECK_FLAG(api->message, ZAPI_MESSAGE_OPAQUE)) {
		STREAM_GETW(s, api->opaque.length);
		assert(api->opaque.length < ZAPI_MESSAGE_OPAQUE_LENGTH);

		STREAM_GET(api->opaque.data, s, api->opaque.length);
	}
```

Reply:
> But re->opaque->data can be anything, not only `struct bgp_zebra_opaque`.

That is the issue: if `re->opaque->data` can be anything it is possible when you'll use it in `zebra` as `struct bgp_zebra_opaque` (for an example) it is possible that it might not be the size you expect. The `memcpy` I've commented could in theory be run with the wrong allocated size.

I see two options:
* Check opaque data size before touching it (potentially saves memory usage)
* Always allocate the biggest opaque type possible (opaque data would not be opaque anymore or a union, potentially saving CPU by not having to do size checks)

Or we could do nothing and I might be paranoid about this.

Reply:
we did have some discussion about this when this opaque business was introduced (by ... Donald). I think the conclusion was that there's a contract between the daemon and zebra; zebra has a callback or handler that does know what is in the "opaque" blob, and knows how to format it for output/display. so it's not paranoid - it's the design...

Reply:
Changed to lib/route_opaque.h.